### PR TITLE
Record link counts in each entry and resolve hardlinks on each lookin…

### PR DESCRIPTION
**Pull request by ktock**
Fixes: #20

Currently we can't deal with hard-link and there are two causes:
- In `TOCEntriy`s, link counts aren't recorded and the filesystem returns a
  static number "1" as link counts.
- When a filesystem looksup to a stargz file (with `stargz.Reader.Lookup()` or
  `(*stargz.TOCEntry).LookupChild()`), the hardlink isn't resolved.

This commit solves this issue by making a TOCEntry field which records each
entry's link counts and by resolving hardlinks on every lookingup.

### CODE REVIEW

File: stargz/stargz.go [hunk 1]
```diff
@@ -175,6 +177,9 @@ type TOCEntry struct {
 	ChunkSize   int64 `json:"chunkSize,omitempty"`
 
 	children map[string]*TOCEntry
+
+	// Nlink is the number of entry names pointing to this entry.
+	Nlink int
```
Comment (bradfitz) [inline stargz/stargz.go]: Put this above the unexported children field.

I'd probably put it between DevMinor and ChunkOffset.

Let's name it NumLink too. And document what zero means. Does zero mean one?

File: stargz/stargz.go [hunk 2]
```diff
@@ -321,6 +326,15 @@ func (r *Reader) initFields() {
 			name = strings.TrimSuffix(name, "/")
 		}
 		pdir := r.getOrCreateDir(parentDir(name))
+		ent.Nlink++ // at least one name(ent.Name) references this entry.
+		if ent.Type == "hardlink" {
+			if org, ok := r.m[ent.LinkName]; ok {
+				org.Nlink++ // original entry is referenced by this ent.Name.
+				ent = org
+			} else {
+				return fmt.Errorf("%s is hardlink but the linkname %s isn't found", ent.Name, ent.LinkName)
```
Comment (bradfitz) [inline stargz/stargz.go]: use %q for strings in errors

Comment (bradfitz) [reply stargz/stargz.go]: and "is a hardlink"

File: stargz/stargz.go [hunk 3]
```diff
@@ -390,6 +406,9 @@ func (r *Reader) Lookup(path string) (e *TOCEntry, ok bool) {
 	// here? And it probably needs a link count field stored in
 	// the TOCEntry.
```
Comment (bradfitz) [inline stargz/stargz.go]: Is this TODO now done?

File: crfs.go [hunk 4]
```diff
@@ -946,7 +946,7 @@ func (n *node) Attr(ctx context.Context, a *fuse.Attr) error {
 	a.Uid = uint32(n.te.Uid)
 	a.Gid = uint32(n.te.Gid)
 	a.Rdev = uint32(unix.Mkdev(uint32(n.te.DevMajor), uint32(n.te.DevMinor)))
-	a.Nlink = 1 // TODO: get this from te once hardlinks are more supported
+	a.Nlink = uint32(n.te.Nlink)
```
Comment (bradfitz) [inline crfs.go]: Should this map 0 to 1?

File: stargz/stargz.go (hunk 1 above)
Comment (ktock) [reply stargz/stargz.go]: Thank you for your review. I fixed it and now NumLink=0 means one name references the entry.

File: stargz/stargz.go (hunk 2 above)
Comment (ktock) [reply stargz/stargz.go]: Thank you for your review. I fixed it.

File: stargz/stargz.go (hunk 3 above)
Comment (ktock) [reply stargz/stargz.go]: Yes, this is done. So I deleted the comment.

File: crfs.go (hunk 4 above)
Comment (ktock) [reply crfs.go]: Yes, now NumLink=0 means one name references the entry so I fixed it.

File: crfs.go [hunk 5]
```diff
@@ -946,7 +946,7 @@ func (n *node) Attr(ctx context.Context, a *fuse.Attr) error {
 	a.Uid = uint32(n.te.Uid)
 	a.Gid = uint32(n.te.Gid)
 	a.Rdev = uint32(unix.Mkdev(uint32(n.te.DevMajor), uint32(n.te.DevMinor)))
-	a.Nlink = 1 // TODO: get this from te once hardlinks are more supported
+	a.Nlink = uint32(n.te.NumLink + 1)
```
Comment (bradfitz) [inline crfs.go]: No, not an unconditional plus one. I meant zero should mean one and one should mean one and two should mean two.

In Go, the zero value should always be reasonable. If we really need to represent zero (perhaps in the future), then we can declare that negative values map to zero.

File: stargz/stargz_test.go [hunk 6]
```diff
@@ -161,9 +161,9 @@ func TestWriteAndOpen(t *testing.T) {
 			),
 			wantNumGz: 3,
 			want: checks(
-				lookupMatch("b", &TOCEntry{Name: "b", Type: "block", DevMajor: 123, DevMinor: 456}),
-				lookupMatch("c", &TOCEntry{Name: "c", Type: "char", DevMajor: 111, DevMinor: 222}),
-				lookupMatch("f", &TOCEntry{Name: "f", Type: "fifo"}),
+				lookupMatch("b", &TOCEntry{Name: "b", Type: "block", DevMajor: 123, DevMinor: 456, Nlink: 1}),
+				lookupMatch("c", &TOCEntry{Name: "c", Type: "char", DevMajor: 111, DevMinor: 222, Nlink: 1}),
+				lookupMatch("f", &TOCEntry{Name: "f", Type: "fifo", Nlink: 1}),
```
Comment (bradfitz) [inline stargz/stargz_test.go]: Did you forget to run the tests? This shouldn't compile?

**Comment by ktock** [comment]
@bradfitz Sorry for not enough fixes. I fixed them.

**Pull request merged**